### PR TITLE
[serve] Refactor `UserCallableWrapper` to return `concurrent.futures.Future`

### PR DESCRIPTION
**Pull request by edoakes**
## Why are these changes needed?

Previously, `UserCallableWrapper` wrapped results in an `asyncio.Future`. This was convenient for use in the replica, but this code will be re-used from a sync context for local testing mode, so I need access to the underlying `concurrent.futures.Future` instead.

Alternatively we could add a duplicate path that returns the concurrent future instead, but I actually prefer that this is more explicit that the code is running on a separate thread (will help defend against mistaken bugs in the future).

I've also moved the `_user_callable_thread` and event loop into instance attributes instead of class attributes. I must have missed something when implementing this previously because this was basically trivial.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by GeneDer — APPROVED** [review]
:shipit:

**Comment by can-anyscale** [comment]
Broke https://github.com/ray-project/ray/issues/47950 and blocking the release; I'm reverting to unblock

**Comment by can-anyscale** [comment]
Reverted PR: https://github.com/ray-project/ray/pull/48468

**Pull request merged**
